Rename Dimname::name to Dimname::full_name

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21795 Implement tensor.select(Dimname,int)
* #21789 Add at::dimname_to_position helper.
* #21752 named inference rule for tensor.select
* #21781 Disallow creation of tensors with duplicate names
* **#21803 Rename Dimname::name to Dimname::full_name**
* #21735 Copy NamedTensorMeta in TensorImpl::copy_tensor_data()
* #21648 Expose torch.empty(sizes, *, names, ...) to Python

This makes it clearer: the name field is the full name (untagged_name + tag).

Test Plan:
- [namedtensor ci]

Differential Revision: [D15833452](https://our.internmc.facebook.com/intern/diff/D15833452)

zou3519 committed:
Rename Dimname::name to Dimname::full_name

This makes it clearer: the name field is the full name (untagged_name + tag).

Test Plan:
- [namedtensor ci]

nairbv commented on aten/src/ATen/Dimname.h:
```diff
@@ -14,16 +14,23 @@ struct CAFFE2_API Dimname {
+  //
+  // For "C.in":
+  // - "C.in" is the "full name"
+  // - "C" is the "name"
```
should this say "C" is the "untagged_name"?

nairbv commented on aten/src/ATen/Dimname.h:
```diff
@@ -14,16 +14,23 @@ struct CAFFE2_API Dimname {
-    : untagged_name_(untagged_name), name_(name), type_(type) {}
+    : untagged_name_(name), full_name_(name), type_(NameType::NORMAL) {}
+  Dimname(NameType type, Symbol full_name, Symbol untagged_name)
+    : untagged_name_(untagged_name), full_name_(full_name), type_(type) {}
```
is there a reason you prefer untagged_name and full_name over name and tag? To me untagged_name and full_name seem repetitive since most of the chars are in both, and then you have to manipulate a string to get the tag on its own. Is the use case just such that the isolated tag is rarely needed?

zou3519 replied:
We never need the isolated tag. It's not repetitive information because we're storing Symbols, not strings.

If we stored only `name` and `tag` it is difficult to get the `full_name` Symbol: we'd need to convert `name` to a std::string, convert `tag` to std::string, add them, and then reconstruct the Symbol. This is all very inefficient.

My hypothesis is that we need to use the `full_name` Symbol much more than the `tag` Symbol (I'm pretty sure a `tag` Symbol would never get used, but I'm not sure if if we actually need the `full_name` Symbol).

nairbv replied:
thanks

zou3519 committed:
Update on "Rename Dimname::name to Dimname::full_name"

Rename Dimname::name to Dimname::full_name

This makes it clearer: the name field is the full name (untagged_name + tag).

Test Plan:
- [namedtensor ci]

gh-metadata: pytorch pytorch 21803 gh/zou3519/55/head
